include/navigation/perfect_memory.h

```diff
@@ -115,7 +103,29 @@ class PerfectMemory : public VisualNavigationBase
     // Private members
     //------------------------------------------------------------------------
     Store m_Store;
-    mutable std::vector<float> m_Differences;
+
+    auto &testInternal(const cv::Mat &image) const
+    {
+        const auto &unwrapRes = getUnwrapResolution();
+        BOB_ASSERT(image.cols == unwrapRes.width);
+        BOB_ASSERT(image.rows == unwrapRes.height);
+        BOB_ASSERT(image.type() == CV_8UC1);
+
+        const size_t numSnapshots = getNumSnapshots();
+        BOB_ASSERT(numSnapshots > 0);
+
+        // Clear differences
+        thread_local std::vector<float> diffs;
+        diffs.reserve(numSnapshots);
+        diffs.clear();
+
+        // Loop through snapshots and caculate differences
+        for (size_t s = 0; s < numSnapshots; s++) {
```

Comment:
shouldn't *this* be the loop you parallelise as the iterations are entirely independent